File: source/necg.f90
!PROJECT TITLE: N-electron correlated gaussians
!
!     AUTHOR: Mauricio Cafiero
!     PROJECT SUMMARY: N-electron correlated gaussians, born-oppenheimer, using shifted centers.
PROGRAM  necg
!SHARED MODULE ***********************************************************************************
USE MATRICES_TO_BE_SHARED



!Declarations ************************************************************************************
IMPLICIT NONE 

include 'mpif.h'


DOUBLE PRECISION:: DP,dp2,dp3,dp4,d1,d2,d3,d4,d5,d6,d7,d8,d9,store,eps,ETA,STEPmx,ACCRCY,XTOL	
double precision, allocatable,dimension(:)::hiv	,w,hiv2,gr,v1,v2,work
INTEGER::I,J,JJ	,l,k,lw,ii,MAXIT,MSGLVL,MAXFUN,kk,stf,lwork,hostnm,stfe	
double precision, allocatable,dimension(:,:)::ea,eb,iden		

CHARACTER (LEN=10) :: TT			
CHARACTER (LEN=5)  :: TZ
CHARACTER (LEN=8)  :: TD
CHARACTER (LEN=80) :: CHARSTRING,recode	

external outprod

call MPI_INIT(ierr)
call MPI_COMM_RANK(MPI_COMM_WORLD,rank, ierr)
call MPI_COMM_SIZE(MPI_COMM_WORLD,nproc,ierr)

!***************************************************************************************************
!Prepare output
  
!!$flag=hostnm(hostname)
!!$write(*,*)'hello from ',hostname

if(rank==0)then

   OPEN (7, FILE = 'necg.out', ACCESS = 'SEQUENTIAL')	

   CALL BIOUT('N-particle Non-Born-Oppenheimer Correlated Gaussians')
   CALL BIOUT('===================================='//'===============')
   CALL BIOUT('ADAMOWICZ RESEARCH                 '  &
  			//'[UofA chemistry]')
   CALL BIOUT('Programming: Mauricio Cafiero ') 
   CALL BIOUT('===================================='  &
  	               //'===============')


  													
   CALL BIOUT(':PROGRAM BEGINS:')				
   CALL Get_Current_Time(TD,TT,TZ)
   CHARSTRING='Date: '//TD//' | Time: '//TT//' / '//TZ//' gmt'
   CALL BIOUT(CHARSTRING)
end if

call read_input
 				
if(rank==0)then
DO I=6,7
  	WRITE(I,*)'M, NN, NE,NST,  =',M,ne, nst
  	WRITE(I,*)'Electric Field  =',efs
  	WRITE(I,*)'OUTPUT FILENAME     =','necg.out'
  
END DO
end if


lwork=3*m-1

Allocate(hiv(siv+m),hiv2(siv+m),ea(m,m),eb(m,m),work(lwork),  &
         v1(svh+skp),v2(svh+skp),gr(siv+m),iden(m,m))


!**************


!!$call matel(1,1,2,d1,d2,d3,d4)
!!$write(*,*)'hello'
!!$d5=d4
!!$eps=0.00001
!!$c(1)=c(1)+eps
!!$iv(91)=iv(91)+eps
!!$call matel(1,1,2,d1,d2,d3,d4)
!!$write(*,*)'hello'
!!$call gradel(1,1,2,v1,v2)
!!$d6=(d4-d5)/eps
!!$write(*,*)'gradient:',d6
!!$stop

!******************************


h=zero
s=zero
kinetic=zero
potential=zero
ki=zero
po=zero
hh=zero
ss=zero

if(rank==0)then 
d5=MPI_WTIME() 
end if       

do i=rank+1,m,nproc
   do j=i,m
      do stf=1, nst
      call matel(stf,j,i,dp,dp2,dp3,dp4)
      kinetic(j,i)=kinetic(j,i)+scvv(stf)*dp2
      potential(j,i)=potential(j,i)+scvv(stf)*(dp3)
      h(j,i)= h(j,i)+scvv(stf)*dp2+scvv(stf)*(dp3)-scvv(stf)*dp4 
      s(j,i)=s(j,i)+scvv(stf)*dp
   end do
end do
end do

call MPI_REDUCE(kinetic,ki,m*m, MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)
call MPI_REDUCE(potential,po,m*m,MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)
call MPI_REDUCE(h,hh,m*m, MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)
call MPI_REDUCE(s,ss,m*m, MPI_DOUBLE_PRECISION,MPI_SUM,0,MPI_COMM_WORLD,ierr)

if(rank==0)then
d6=MPI_WTIME()
d7=d6-d5
write(*,*)'time for matrix calc. is',d7,'seconds'
call flush(6)
end if
!!$goto 97

if(rank==0)then
s=ss
h=hh
kinetic=ki
potential=po
end if
!!$!stop

if(rank==0)then
ea=h
eb=s
!****
!!$iden=zero
!!$do i=1,m
!!$   iden(i,i)=one
!!$end do
!!$allocate(w(m))
!!$call DSYGV(1,'V','L',m,eb,m,iden,m,w,work,lwork,jj)
!!$write(*,*)'error from eigenvalue:',jj
!!$write(*,*)'Eigenvalue=',w
!!$deallocate(w,work)
!!$do i=1,m
!!$   write(*,*)eb(i,1)
!!$end do
!!$end if
!!$stop
!!$if(rank==0)then
!****
do j=1,m
   do i=1,j
      h(i,j)=h(j,i)
      s(i,j)=s(j,i)
      kinetic(i,j)=kinetic(j,i)
      potential(i,j)=potential(j,i)
   end do
end do

!!$!call tab(s,m,m,m,m)
!stop
!goto 16
allocate(w(m))
call DSYGV(1,'V','L',m,ea,m,eb,m,w,work,lwork,jj)
write(*,*)'error from eigenvalue:',jj
write(*,*)'Eigenvalue=',w(1)
do i=1,m
   c(i)=ea(i,1)
end do
deallocate(w,work)

16 j=1
do i=1,siv
   hiv(i)=iv(i)
   j=j+1
end do
do i=1,m
   hiv(j)=c(i)
   j=j+1
end do

DO I=6,7
  	WRITE(I,*)'********** The Input Vectors ****************************************'
  	WRITE(I,*)''
END DO
!write(*,*)iv
!write(*,*)''
write(*,*)c
DO I=6,7
  	WRITE(I,*)'********** End Input Vectors ****************************************'
  	WRITE(I,*)''
END DO

d1=zero
d2=zero
d4=zero
d5=zero
do i=1,m
   do j=1,m
      d1=d1+c(i)*h(i,j)*c(j)
      d2=d2+(c(i)*s(i,j)*c(j))
      d4=d4+c(i)*kinetic(i,j)*c(j)
      d5=d5+c(i)*potential(i,j)*c(j)
   end do
end do

d3=d1/d2
d4=d4/d2
d5=d5/d2
d6=-two*d4/d5

write(*,*)'*********************************************************************'
write(*,*)''
write(*,*)'Rayleigh Quotient: ', d3
write(*,*)''
write(*,*)'Average Kinetic  : ', d4
write(*,*)''
write(*,*)'Average Potential: ', d5
write(*,*)''
write(*,*)'Virial Coef.     : ', d6
write(*,*)''
write(*,*)'*********************************************************************'

end if
if(rank==0)then
d5=MPI_WTIME()
end if
call outprod(siv+m,hiv,d1,gr)
if(rank==0)then
d6=MPI_WTIME()
d7=d6-d5
write(*,*)'time for energy/gradient calc. is',d7,'seconds'
write(*,*)'energy is ',d1,'hartree'
call flush(6)
end if
call MPI_BCAST(hiv,siv+m,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)
call MPI_BCAST(d3,1,MPI_DOUBLE_PRECISION,0,MPI_COMM_WORLD,ierr)

!****************  numerical optimization ***********
!l=siv+m
!lw=l*(l+10)
!lw=lw+10
!write(*,*)'lw:',lw
!allocate(w(lw))
!write(*,*)'hello'
!call uncmnd(l,hiv,outprod,hiv2,d3,jj,w,lw)

!j=1
!do i=1,siv
!   iv(i)=hiv2(i)
!   j=j+1
!end do
!do i=1,m
!   c(i)=hiv2(j)
!   j=j+1
!end do
!goto 23
!****************************************************

!***************** analytical optimization **********

!Set up parameters for the optimization program
      MAXIT = 20
      MSGLVL = 1
      MAXFUN = 150*(siv+m)
      ETA = 0.21D0
      STEPMX = 0.7D1
      ACCRCY = 2.220446649250313D-16
      XTOL = DSQRT(ACCRCY)
      l=(siv+m)
      lw=14*l
      allocate(w(lw))

CALL LMQN(jj,l,hiv,d3,Gr,W,LW,outprod,MSGLVL,MAXIT,MAXFUN,ETA,STEPMX,ACCRCY,XTOL,rank)
  

if(rank==0)then

wRITE(*,*)"IERROR",jj
j=1
do i=1,siv
   iv(i)=hiv(i)
   j=j+1
end do
do i=1,m
   c(i)=hiv(j)
   j=j+1
end do
!end if
!***********************************************************************

23 DO I=6,7
  	WRITE(I,*)'********** The Output Vectors ***************************************'
  	WRITE(I,*)''
END DO
!write(*,*)iv
!write(*,*)''
write(*,*)c
write(*,*)''
DO I=6,7
  	WRITE(I,*)'********** End Output Vectors ***************************************'
  	WRITE(I,*)''
END DO

!DO I=6,7
!  	WRITE(I,*)'********** The Final Gradient Vector ********************************'
!  	WRITE(I,*)''
!END DO
!write(*,*)gr
!write(*,*)''
!DO I=6,7
! 	WRITE(I,*)'*********************************************************************'
!  	WRITE(I,*)''
!END DO

!call tab (H,m,m,m,m)
h=zero
s=zero
kinetic=zero
potential=zero
do stf=1,nst
!stf=1
do i=1,m
   do j=i,m
      call matel(stf,j,i,dp,dp2,dp3,dp4)
      kinetic(j,i)=kinetic(j,i)+scvv(stf)*dp2
      potential(j,i)=potential(j,i)+scvv(stf)*(dp3)
      h(j,i)= h(j,i)+scvv(stf)*dp2+scvv(stf)*(dp3)-scvv(stf)*dp4
      s(j,i)=s(j,i)+scvv(stf)*dp
   end do
end do
end do


!!$!stop
do j=1,m
   do i=1,j
      h(i,j)=h(j,i)
      s(i,j)=s(j,i)
      kinetic(i,j)=kinetic(j,i)
      potential(i,j)=potential(j,i)
   end do
end do

d1=zero
d2=zero
d4=zero
d5=zero
do i=1,m
   do j=1,m
      d1=d1+c(i)*h(i,j)*c(j)
      d2=d2+(c(i)*s(i,j)*c(j))
      d4=d4+c(i)*kinetic(i,j)*c(j)
      d5=d5+c(i)*potential(i,j)*c(j)
   end do
end do

d3=d1/d2
d4=d4/d2
d5=d5/d2
d6=-two*d4/d5
!!$
!!$
write(*,*)'********************************'
write(*,*)''
write(*,*)'Rayleigh Quotient: ', d3
write(*,*)''
write(*,*)'Average Kinetic  : ', d4
write(*,*)''
write(*,*)'Average Potential: ', d5
write(*,*)''
write(*,*)'Virial Coef.     : ', d6
write(*,*)''
write(*,*)'********************************'

d9=zero
d7=zero
d8=zero
kk=1
do ii=1,ne
do jj=1,3


h=zero
s=zero

do stfe=1,nst
do stf=1,nst
!stf=1
do i=1,m
   do j=i,m

      call rrms(stfe,stf,j,i,kk,dp,dp2)
      h(j,i)= h(j,i)+scvv(stf)*scvv(stfe)*dp2
      s(j,i)=s(j,i)+scvv(stf)*scvv(stfe)*dp
   end do
end do
end do
end do

do j=1,m
   do i=1,j
      h(i,j)=h(j,i)
      s(i,j)=s(j,i)
   end do
end do
!!$call biout('Hamiltonian Matrix')
!!$call tab(h,m,m,m,m)
!!$call biout('Overlap Matrix')
!!$call tab(s,m,m,m,m)
!!$stop


d1=zero
d2=zero
do i=1,m
   do j=1,m
      d1=d1+c(i)*h(i,j)*c(j)
      d2=d2+(c(i)*s(i,j)*c(j))
   end do
end do
!!$write(*,*)d1,d2
d3=d1/d2

if(jj==1)then
   d7=d7+ncv(ii+1)*d3
else if(jj==2)then
   d8=d8+ncv(ii+1)*d3
else if(jj==3)then
   d9=d9+ncv(ii+1)*d3
end if


kk=kk+1
end do
end do

d4=d7*2.541765d+000
d5=d8*2.541765d+000
d6=d9*2.541765d+000

write(*,*)''
write(*,*)'Dipole x-component=', d7,'(',d4,')'
write(*,*)''

write(*,*)''
write(*,*)'Dipole y-component=', d8,'(',d5,')'
write(*,*)''

write(*,*)''
write(*,*)'Dipole z-component=', d9,'(',d6,')'
write(*,*)''

write(*,*)'*********************************************************************'

  CALL BIOUT('PROGRAM TERMINATES')
  CALL Get_Current_Time(TD,TT,TZ)
  CHARSTRING='Date: '//TD//' | Time: '//TT//' / '//TZ//' gmt'
  CALL BIOUT(CHARSTRING)

end if

call MPI_FINALIZE(ierr)

end program necg
 

File: python/necg.py
"""
N-electron Correlated Gaussians (NECG) — Python implementation.

Project title : N-particle Non-Born-Oppenheimer Correlated Gaussians
Original code : ADAMOWICZ RESEARCH [UofA chemistry]
Original prog : Mauricio Cafiero
Python port   : See repository README

Analogue of necg.f90.  MPI parallelism has been removed; numpy / scipy
are used for all linear-algebra operations.

Usage
-----
    python necg.py [input_file]

The input file defaults to ``oldquit.dat`` (matching the Fortran version).
Output is written to stdout **and** to ``necg.out``.
"""

from __future__ import annotations

import sys

import numpy as np
from scipy import linalg
from scipy.optimize import minimize

import shared as sh
from read_input import read_input
from matel import matel
from outprod import outprod
from rrms import rrms
from utils import biout, get_current_time


def _build_matrices() -> None:
    """Fill H, S, Kinetic and Potential from scratch using current iv / C."""
    sh.H[:] = 0.0
    sh.S[:] = 0.0
    sh.Kinetic[:] = 0.0
    sh.Potential[:] = 0.0

    for i in range(1, sh.M + 1):
        for j in range(i, sh.M + 1):
            for stf in range(1, sh.nst + 1):
                ov, tk, ert, zht = matel(stf, j, i)
                sh.Kinetic[j - 1, i - 1] += sh.scvv[stf - 1] * tk
                sh.Potential[j - 1, i - 1] += sh.scvv[stf - 1] * ert
                sh.H[j - 1, i - 1] += sh.scvv[stf - 1] * (tk + ert - zht)
                sh.S[j - 1, i - 1] += sh.scvv[stf - 1] * ov

    # Symmetrise
    for j in range(sh.M):
        for i in range(j):
            sh.H[i, j] = sh.H[j, i]
            sh.S[i, j] = sh.S[j, i]
            sh.Kinetic[i, j] = sh.Kinetic[j, i]
            sh.Potential[i, j] = sh.Potential[j, i]


def _rayleigh_quotient() -> tuple[float, float, float, float]:
    """Return (energy, kinetic, potential, virial) using current C, H, S."""
    cHc = float(sh.C @ sh.H @ sh.C)
    cSc = float(sh.C @ sh.S @ sh.C)
    cKc = float(sh.C @ sh.Kinetic @ sh.C)
    cPc = float(sh.C @ sh.Potential @ sh.C)
    energy = cHc / cSc
    kin = cKc / cSc
    pot = cPc / cSc
    virial = -2.0 * kin / pot
    return energy, kin, pot, virial


def _print_energetics(energy: float, kin: float, pot: float, virial: float) -> None:
    print("*" * 69)
    print()
    print(f"Rayleigh Quotient: {energy:25.15f}")
    print()
    print(f"Average Kinetic  : {kin:25.15f}")
    print()
    print(f"Average Potential: {pot:25.15f}")
    print()
    print(f"Virial Coef.     : {virial:25.15f}")
    print()
    print("*" * 69)


def main(input_file: str = "oldquit.dat") -> None:
    """Main driver — mirrors the flow of necg.f90."""

    # ------------------------------------------------------------------
    # Initialise output
    # ------------------------------------------------------------------
    import os
    if os.path.exists("necg.out"):
        os.remove("necg.out")

    biout("N-particle Non-Born-Oppenheimer Correlated Gaussians")
    biout("=" * 68)
    biout("ADAMOWICZ RESEARCH                  [UofA chemistry]")
    biout("Programming: Mauricio Cafiero")
    biout("=" * 68)
    biout(":PROGRAM BEGINS:")
    td, tt, tz = get_current_time()
    biout(f"Date: {td} | Time: {tt} / {tz} gmt")

    # ------------------------------------------------------------------
    # Read input
    # ------------------------------------------------------------------
    read_input(input_file)

    print(f"M, NE, NST = {sh.M}, {sh.ne}, {sh.nst}")
    print(f"Electric Field = {sh.efs}")
    print("OUTPUT FILENAME = necg.out")

    # ------------------------------------------------------------------
    # Build initial H and S matrices
    # ------------------------------------------------------------------
    import time
    t0 = time.time()
    _build_matrices()
    t1 = time.time()
    print(f"Time for matrix calc. is {t1 - t0:.3f} seconds")

    # ------------------------------------------------------------------
    # Solve generalised eigenvalue problem  H·v = λ·S·v
    # to initialise C (scipy.linalg.eigh replaces DSYGV)
    # ------------------------------------------------------------------
    try:
        eigenvalues, eigenvectors = linalg.eigh(sh.H, sh.S)
        print(f"Eigenvalue = {eigenvalues[0]:.15f}")
        sh.C[:] = eigenvectors[:, 0]
    except linalg.LinAlgError as exc:
        print(f"Warning: generalised eigenvalue problem failed: {exc}")

    print()
    print("***** The Input Vectors *****")
    print(sh.C)
    print("***** End Input Vectors *****")
    print()

    # ------------------------------------------------------------------
    # Initial energetics
    # ------------------------------------------------------------------
    energy, kin, pot, virial = _rayleigh_quotient()
    _print_energetics(energy, kin, pot, virial)

    # ------------------------------------------------------------------
    # Assemble initial parameter vector for optimiser
    # ------------------------------------------------------------------
    hiv = np.concatenate([sh.iv.copy(), sh.C.copy()])

    # ------------------------------------------------------------------
    # Optimisation using L-BFGS-B (scipy analogue of LMQN)
    # scipy.optimize.minimize(method="L-BFGS-B") provides a limited-
    # memory quasi-Newton method analogous to the LMQN Fortran routine.
    # ------------------------------------------------------------------
    t0 = time.time()
    result = minimize(
        outprod,
        hiv,
        method="L-BFGS-B",
        jac=True,                           # outprod returns (f, g)
        options={
            "maxiter": 20,
            "maxfun": 150 * (sh.siv + sh.M),
            "ftol": 2.220446649250313e-16,
            "gtol": np.sqrt(2.220446649250313e-16),
            "iprint": 1,
        },
    )
    t1 = time.time()
    print(f"Time for energy/gradient calc. is {t1 - t0:.3f} seconds")
    print(f"Energy is {result.fun:.15f} hartree")
    print(f"Optimiser message: {result.message}")

    # ------------------------------------------------------------------
    # Unpack optimised parameters
    # ------------------------------------------------------------------
    sh.iv[:] = result.x[:sh.siv]
    sh.C[:] = result.x[sh.siv:]

    print()
    print("***** The Output Vectors *****")
    print(sh.C)
    print()
    print("***** End Output Vectors *****")
    print()

    # ------------------------------------------------------------------
    # Final matrix build with optimised parameters
    # ------------------------------------------------------------------
    _build_matrices()
    energy, kin, pot, virial = _rayleigh_quotient()
    _print_energetics(energy, kin, pot, virial)

    # ------------------------------------------------------------------
    # Dipole moment calculation (replaces the rrms loop in necg.f90)
    # ------------------------------------------------------------------
    d7 = d8 = d9 = 0.0
    kk = 1
    for ii in range(1, sh.ne + 1):
        for jj_comp in range(1, 4):
            h_dip = np.zeros((sh.M, sh.M))
            s_dip = np.zeros((sh.M, sh.M))

            for stfe in range(1, sh.nst + 1):
                for stf in range(1, sh.nst + 1):
                    for i in range(1, sh.M + 1):
                        for j in range(i, sh.M + 1):
                            ov_r, tk_r = rrms(stfe, stf, j, i, kk)
                            h_dip[j - 1, i - 1] += sh.scvv[stf - 1] * sh.scvv[stfe - 1] * tk_r
                            s_dip[j - 1, i - 1] += sh.scvv[stf - 1] * sh.scvv[stfe - 1] * ov_r

            # Symmetrise
            for j in range(sh.M):
                for i in range(j):
                    h_dip[i, j] = h_dip[j, i]
                    s_dip[i, j] = s_dip[j, i]

            cHc_d = float(sh.C @ h_dip @ sh.C)
            cSc_d = float(sh.C @ s_dip @ sh.C)
            d3 = cHc_d / cSc_d if cSc_d != 0.0 else 0.0

            if jj_comp == 1:
                d7 += sh.ncv[ii] * d3
            elif jj_comp == 2:
                d8 += sh.ncv[ii] * d3
            else:
                d9 += sh.ncv[ii] * d3

            kk += 1

    DEBYE = 2.541765e0
    d4 = d7 * DEBYE
    d5 = d8 * DEBYE
    d6 = d9 * DEBYE

    print()
    print(f"Dipole x-component = {d7:.15f}  ( {d4:.6f} Debye )")
    print()
    print(f"Dipole y-component = {d8:.15f}  ( {d5:.6f} Debye )")
    print()
    print(f"Dipole z-component = {d9:.15f}  ( {d6:.6f} Debye )")
    print()
    print("*" * 69)

    biout("PROGRAM TERMINATES")
    td, tt, tz = get_current_time()
    biout(f"Date: {td} | Time: {tt} / {tz} gmt")


if __name__ == "__main__":
    input_file = sys.argv[1] if len(sys.argv) > 1 else "oldquit.dat"
    main(input_file)
